test/TestPeps.m

```diff
@@ -0,0 +1,197 @@
+            v3 = apply(mpo, v2);
+        end
+        
+        function testDerivatives(tc, spaces, depth, width, dualdepth, dualwidth, samebot)
+            tc.assumeTrue(braidingstyle(spaces) ~= BraidingStyle.Fermionic, 'fZ2 test broken')
+
+            mpo = tc.random_inf_mpo(spaces, depth, width, dualdepth, dualwidth, samebot);
+            vspaces = repmat({spaces(end)}, 1, width);
+            mps = mpo.initialize_mps(vspaces{:});
+            
+            [GL, GR] = environments(mpo, mps, mps);
+            
+            H_AC = AC_hamiltonian(mpo, mps, GL, GR);
+            for i = 1:numel(H_AC)
+                AC_ = mps.AC(i);
+                [AC_.var, lambda] = eigsolve(H_AC{i}, mps.AC(i).var, 1, 'largestabs');
+                AC_2 = apply(H_AC{i}, AC_);
+                tc.assertTrue(isapprox(AC_2, lambda * AC_.var, 'RelTol', 1e-6));
+            end
+            
+            H_C = C_hamiltonian(mpo, mps, GL, GR);
+            for i = 1:numel(H_C)
+                [C_, lambda] = eigsolve(H_C{i}, mps.C(i), 1, 'largestabs');
+                tc.assertTrue(isapprox(apply(H_C{i}, C_), lambda * C_, 'RelTol', 1e-6));
+            end
+        end
+        
+        function testMpoTensor(tc, spaces, dualdepth, dualwidth)
+            tc.assumeTrue(braidingstyle(spaces) ~= BraidingStyle.Fermionic, 'fZ2 test broken')
+
+            pspace = spaces(1); horzspace = spaces(2); vertspace = spaces(3);
+            top = TestPeps.random_peps(pspace, horzspace, vertspace, 1, 1, dualdepth, dualwidth);
+            bot = TestPeps.random_peps(pspace', vertspace, horzspace, 1, 1, dualdepth, dualwidth);
+            O = PepsSandwich(top{1}, bot{1});
+            t = MpoTensor(O);
+        end
+        
+end
+    
+    methods (Static)
+        function A = random_peps(pspace, horzspace, vertspace, depth, width, dualdepth, dualwidth)
+            % spit out random peps unit cell
+            A = cell(depth, width);
+            if dualwidth, horzspace = horzspace'; end
+            if dualdepth, vertspace = vertspace'; end
+
+            if mod(depth, 2)
+                vertspaces = [vertspace vertspace'];
+            else
+                vertspaces = [vertspace vertspace];
+            end
+
+            if mod(width, 2)
+                horzspaces = [horzspace horzspace'];
+            else
+                horzspaces = [horzspace horzspace];
+            end
+
+            codomainspaces = reshape([vertspaces; horzspaces], 1, []);
+
+            for d = 1:depth
+                for w = 1:width
+                    dmsp = pspace;
+                    cdmsp = codomainspaces;
+                    if ~mod(depth, 2) && ~mod(d, 2), cdmsp = conj(cdmsp); dmsp = conj(dmsp); end
+                    if ~mod(width, 2) && ~mod(w, 2), cdmsp = conj(cdmsp); dmsp = conj(dmsp); end
+                    A{d, w} = Tensor.randnc(dmsp, cdmsp);
+                    A{d, w} = A{d, w} / norm(A{d, w});
+                end
+            end
+        end
```

Comment:
misschien eerder dit naar `PepsTensor`
`A = PepsTensor.new(fun, pspace, westvspace, southvspace, eastvspace, northvspace)`

eventueel in de toekomst een `InfPeps.new(fun, pspaces, horizontalspaces, verticalspaces)`?